Fix file.managed on Windows with test=True

twangboy opened the pull request:
### What does this PR do?
Fix file.managed with makedirs=True on Windows when the target directory structure is missing and test=True

Regression caused by https://github.com/saltstack/salt/pull/39773

Fix formatting issues in the docs

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/43328

### Tests written?
No

twangboy committed:
Return empty or unmodified dict on file not found

Fixes issue with file.managed with test=True when the directory
structure for the file is not there and makedirs=True

twangboy committed:
Fix documentation formatting

salt-jenkins commented:
@twangboy, thanks for your PR! By analyzing the history of the files in this pull request, we identified @UtahDave, @slai and @Geo23589 to be potential reviewers.

rallytime commented:
@twangboy It would probably be good to add a quick unit test for this so the regression doesn't happen again.

twangboy committed:
Add tests to avoid future regression

twangboy committed:
Remove loader module mixin, add linux paths

twangboy commented:
@rallytime Added some unit tests...

cachedout commented on salt/modules/win_file.py:
```diff
@@ -818,8 +818,10 @@ def stats(path, hash_type='sha256', follow_symlinks=True):
         salt '*' file.stats /etc/passwd
     '''
+    # This is to mirror the behavior of file.py. `check_file_meta` expects an
+    # empty dictionary when the file does not exist
```
I tend to wonder if this should be changed in the other direction. Wouldn't it be cleaner to raise a specific error and then check for it and display it to the user, versus just an empty dict?

twangboy replied:
@cachedout Changed this back, added try/except block to calling function

twangboy committed:
Raise CommandExecutionError when file doesn't exist

Re-add CommandExecutionError in win_file
Add CommandExecutionError in file
Add Try/Except block in file when getting file metadata

twangboy committed:
Fix unit test to handle new Exception

cachedout commented:
It would appear that all the file tests are now broken.

twangboy commented:
Damn....

twangboy committed:
Fix some integration tests

twangboy committed:
Fix some more integration tests for Linux

twangboy committed:
Use file functions for symlink and remove

twangboy committed:
Fix typo

twangboy committed:
Fix integration tests for Windows

twangboy committed:
Fix 2 more tests

twangboy committed:
Add back the try/finally blocks

twangboy commented:
@rallytime I don't know why the GPG test is failing. It looks to me like all the commits are signed...

cachedout commented:
Given that this is a sensitive area and that it's a regression fix, I would like to see additional tests added here to guide future development before we review this.

twangboy commented:
@cachedout There are unit tests... what additional tests do you think we should add here?

twangboy commented:
Rebased...

cachedout commented:
Looked at tests in place. They look sufficient.

rallytime commented:
@twangboy There are two related tests here failing on the python 3 side. Can you take a look?

twangboy commented:
Rebased...

twangboy committed:
Fix test_directory

rallytime commented:
The Python 3 tests here are still failing. They're related to this change. @twangboy Can you take another look? :)

cachedout commented:
Bump @twangboy

twangboy committed:
Fix py3 error

twangboy commented:
Rebased...

rallytime commented:
re-run py2

twangboy commented:
Rebased...

twangboy commented:
Rebased...

twangboy commented:
Rebased...

garethgreenaway committed:
Merge branch '2017.7' into win_fix_file.managed
